File: pilotforce-get-bookings.mjs
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, QueryCommand, ScanCommand } from '@aws-sdk/lib-dynamodb';

// Initialize DynamoDB clients
const client = new DynamoDBClient({ region: 'eu-north-1' });
const docClient = DynamoDBDocumentClient.from(client);
const BOOKINGS_TABLE = 'Bookings';

/**
 * Decode JWT token to extract payload
 * @param {string} token - JWT token
 * @returns {Object|null} - Decoded token payload or null if invalid
 */
function decodeToken(token) {
  try {
    // Split the JWT token into parts
    const parts = token.split('.');
    if (parts.length !== 3) {
      console.error('Token does not have three parts');
      return null;
    }
    
    // Decode the payload (middle part)
    const payload = parts[1];
    const base64 = payload.replace(/-/g, '+').replace(/_/g, '/');
    const jsonPayload = Buffer.from(base64, 'base64').toString('utf8');
    
    return JSON.parse(jsonPayload);
  } catch (error) {
    console.error('Error decoding token:', error);
    return null;
  }
}

export const handler = async (event) => {
  console.log('Event received:', JSON.stringify(event, null, 2));
  
  // CORS headers
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token',
    'Access-Control-Allow-Methods': 'OPTIONS,GET,POST'
  };
  
  // Handle CORS preflight requests
  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({ message: 'CORS preflight successful' })
    };
  }
  
  try {
    // Debug headers received
    console.log('Headers received:', JSON.stringify(event.headers || {}, null, 2));
    
    // Extract token from Authorization header - check both camel case and lowercase
    const authHeader = event.headers?.Authorization || event.headers?.authorization;
    
    console.log('Auth header:', authHeader);
    
    if (!authHeader) {
      console.log('No Authorization header found in request');
      return {
        statusCode: 401,
        headers,
        body: JSON.stringify({ message: 'Authorization token required' })
      };
    }
    
    // Split token from 'Bearer ' prefix if present
    const token = authHeader.startsWith('Bearer ') ? authHeader.substring(7) : authHeader;
    console.log('Token extracted:', token ? `${token.substring(0, 20)}...` : 'null');
    
    // Decode token payload
    const tokenPayload = decodeToken(token);
    
    if (!tokenPayload) {
      console.log('Failed to decode token payload');
      return {
        statusCode: 401,
        headers,
        body: JSON.stringify({ message: 'Invalid token' })
      };
    }
    
    // Extract user info from token
    const userId = tokenPayload.sub;
    let companyId = tokenPayload['custom:companyId'];
    const userRole = tokenPayload['custom:userRole'] || 'User';
    
    console.log(`User ID: ${userId}, Company ID: ${companyId}, Role: ${userRole}`);
    
    // Check for company ID in query params - these override token values
    const queryParams = event.queryStringParameters || {};
    if (queryParams.companyId) {
      companyId = queryParams.companyId;
      console.log(`Using company ID from query params: ${companyId}`);
    }
    
    if (!companyId) {
      console.log('No company ID found in token or query parameters');
      return {
        statusCode: 403,
        headers,
        body: JSON.stringify({ message: 'Company ID not found in token or query parameters' })
      };
    }
    
    // Query using the GSI
    console.log(`Querying bookings for Company ID: ${companyId}`);
    
    try {
      // Try using GSI first
      const queryCommand = new QueryCommand({
        TableName: BOOKINGS_TABLE,
        IndexName: 'CompanyIdIndex',
        KeyConditionExpression: 'CompanyId = :companyId',
        ExpressionAttributeValues: {
          ':companyId': companyId
        }
      });
      
      // Log the actual query parameters for debugging
      console.log('DynamoDB Query parameters:', JSON.stringify({
        TableName: BOOKINGS_TABLE,
        IndexName: 'CompanyIdIndex',
        KeyConditionExpression: 'CompanyId = :companyId',
        ExpressionAttributeValues: {
          ':companyId': companyId
        }
      }));
      
      const queryResponse = await docClient.send(queryCommand);
      let bookings = queryResponse.Items || [];
      
      console.log(`Found ${bookings.length} bookings using CompanyIdIndex`);
      
      // If no results, try with lowercase companyId
      if (bookings.length === 0) {
        console.log('No results from GSI with CompanyId, trying with companyId (lowercase)');
        
        const lowercaseQueryCommand = new QueryCommand({
          TableName: BOOKINGS_TABLE,
          IndexName: 'companyIdIndex', // Try lowercase index name
          KeyConditionExpression: 'companyId = :companyId',
          ExpressionAttributeValues: {
            ':companyId': companyId
          }
        });
        
        try {
          const lowercaseResponse = await docClient.send(lowercaseQueryCommand);
          bookings = lowercaseResponse.Items || [];
          console.log(`Found ${bookings.length} bookings using lowercase companyIdIndex`);
        } catch (lowercaseError) {
          console.log('Error with lowercase index query:', lowercaseError.message);
        }
      }
      
      // If still no results, try scanning with filter
      if (bookings.length === 0) {
        console.log('No results from GSI queries, trying scan with filter');
        
        const scanCommand = new ScanCommand({
          TableName: BOOKINGS_TABLE,
          FilterExpression: 'CompanyId = :companyId OR companyId = :companyId',
          ExpressionAttributeValues: {
            ':companyId': companyId
          }
        });
        
        const scanResponse = await docClient.send(scanCommand);
        bookings = scanResponse.Items || [];
        
        console.log(`Found ${bookings.length} bookings using scan with filter`);
      }
      
      // Log a sample booking for debugging
      if (bookings.length > 0) {
        console.log('Sample booking:', JSON.stringify(bookings[0], null, 2));
      } else {
        console.log('No bookings found after all query attempts');
      }
      
      // Return success with bookings array, even if empty
      return {
        statusCode: 200,
        headers,
        body: JSON.stringify({
          message: bookings.length > 0 ? 'Bookings retrieved successfully' : 'No bookings found',
          bookings: bookings
        })
      };
      
    } catch (queryError) {
      console.error('Error querying DynamoDB:', queryError);
      
      // Always try fallback scan as last resort
      try {
        console.log('Attempting fallback scan for all bookings');
        const scanCommand = new ScanCommand({
          TableName: BOOKINGS_TABLE
        });
        
        const scanResponse = await docClient.send(scanCommand);
        const allBookings = scanResponse.Items || [];
        
        console.log(`Fallback scan found ${allBookings.length} total bookings`);
        
        // Filter client-side by company ID
        const filteredBookings = allBookings.filter(booking => 
          booking.CompanyId === companyId || booking.companyId === companyId
        );
        
        console.log(`Filtered to ${filteredBookings.length} bookings matching company ID`);
        
        return {
          statusCode: 200,
          headers,
          body: JSON.stringify({
            message: 'Bookings retrieved using fallback method',
            bookings: filteredBookings
          })
        };
      } catch (fallbackError) {
        console.error('Fallback scan also failed:', fallbackError);
        throw fallbackError; // Let the outer catch handle this
      }
    }
    
  } catch (error) {
    console.error('Error processing request:', error);
    
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ 
        message: `Internal server error: ${error.message}`,
        error: error.stack
      })
    };
  }
};

File: pilotforce-get-bookings.py
import json
import boto3
import logging
import base64
import os
from boto3.dynamodb.conditions import Key, Attr

# Set up logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)

# Create DynamoDB client
dynamodb = boto3.resource('dynamodb')
bookings_table = dynamodb.Table('Bookings')

def decode_token(token):
    """Decode the JWT token without verification to extract claims"""
    try:
        # For ID tokens, we can decode them without verification to extract user info
        # Split the token into parts
        parts = token.split('.')
        if len(parts) != 3:
            logger.error("Token does not have three parts")
            return None
        
        # Decode the payload (middle part)
        payload_base64 = parts[1]
        # Add padding if needed
        payload_base64 += '=' * ((4 - len(payload_base64) % 4) % 4)
        # Convert from url-safe base64 to standard base64
        payload_base64 = payload_base64.replace('-', '+').replace('_', '/')
        # Decode the base64 string
        payload_json = base64.b64decode(payload_base64)
        # Parse the JSON
        payload = json.loads(payload_json)
        logger.info(f"Decoded token payload: {json.dumps(payload)}")
        return payload
    except Exception as e:
        logger.error(f"Error decoding token: {str(e)}")
        return None

def lambda_handler(event, context):
    # Set up CORS headers
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token',
        'Access-Control-Allow-Methods': 'OPTIONS,GET,POST,DELETE'
    }
    
    # Log the incoming event for debugging
    logger.info(f"Event received: {json.dumps(event)}")
    
    # Handle OPTIONS requests (CORS preflight)
    if event.get('httpMethod') == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({'message': 'CORS preflight successful'})
        }
    
    try:
        # Check if a specific booking ID is requested in the path
        path_parameters = event.get('pathParameters', {}) or {}
        booking_id = path_parameters.get('id')
        
        if booking_id:
            logger.info(f"Fetching specific booking with ID: {booking_id}")
            
            # Query for the specific booking
            response = bookings_table.scan(
                FilterExpression=Attr('BookingId').eq(booking_id) | Attr('id').eq(booking_id)
            )
            
            items = response.get('Items', [])
            
            if not items:
                logger.warning(f"No booking found with ID: {booking_id}")
                return {
                    'statusCode': 404,
                    'headers': headers,
                    'body': json.dumps({'message': f'Booking with ID {booking_id} not found'})
                }
            
            logger.info(f"Found booking: {json.dumps(items[0])}")
            return {
                'statusCode': 200,
                'headers': headers,
                'body': json.dumps(items[0])
            }
        
        # For non-specific requests, check for query parameters
        query_params = event.get('queryStringParameters', {}) or {}
        company_id = query_params.get('companyId')
        
        # If no companyId in query params, try to get it from the token
        if not company_id:
            # Extract authentication token
            auth_header = event.get('headers', {}).get('Authorization') or event.get('headers', {}).get('authorization')
            
            if auth_header:
                token = auth_header.replace('Bearer ', '')
                token_claims = decode_token(token)
                
                if token_claims:
                    company_id = token_claims.get('custom:companyId')
                    logger.info(f"Extracted company ID from token: {company_id}")
        
        # If we have a company ID, filter by it
        if company_id:
            logger.info(f"Fetching bookings for company ID: {company_id}")
            
            # First try with GSI if available
            try:
                response = bookings_table.query(
                    IndexName='CompanyIdIndex',
                    KeyConditionExpression=Key('CompanyId').eq(company_id)
                )
                
                bookings = response.get('Items', [])
                logger.info(f"Found {len(bookings)} bookings via CompanyIdIndex")
                
                # If no results, try with scan fallback
                if not bookings:
                    response = bookings_table.scan(
                        FilterExpression=Attr('CompanyId').eq(company_id) | Attr('companyId').eq(company_id)
                    )
                    bookings = response.get('Items', [])
                    logger.info(f"Found {len(bookings)} bookings via scan filter")
            except Exception as e:
                logger.error(f"Error using GSI: {str(e)}")
                # Fallback to scan
                response = bookings_table.scan(
                    FilterExpression=Attr('CompanyId').eq(company_id) | Attr('companyId').eq(company_id)
                )
                bookings = response.get('Items', [])
                logger.info(f"Found {len(bookings)} bookings via fallback scan")
            
            return {
                'statusCode': 200,
                'headers': headers,
                'body': json.dumps(bookings)
            }
        
        # If no company ID, return all bookings (with a limit for safety)
        logger.info("No company ID provided, fetching all bookings with limit")
        response = bookings_table.scan(Limit=100)
        bookings = response.get('Items', [])
        
        logger.info(f"Returning {len(bookings)} bookings")
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps(bookings)
        }
    
    except Exception as e:
        logger.error(f"Error processing request: {str(e)}")
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'message': f'Internal server error: {str(e)}'})
        }
